universal-application-tool-0.0.1/app/controllers/admin/QuestionController.java

```diff
@@ -0,0 +1,109 @@
+package controllers.admin;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import forms.QuestionForm;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+import play.data.Form;
+import play.data.FormFactory;
+import play.mvc.Controller;
+import play.mvc.Http.Request;
+import play.mvc.Result;
+import services.question.InvalidPathException;
+import services.question.QuestionDefinition;
+import services.question.QuestionService;
+import services.question.UnsupportedQuestionTypeException;
+import views.admin.questions.QuestionEditView;
+import views.admin.questions.QuestionsListView;
+
+public class QuestionController extends Controller {
+
+  private final QuestionService service;
+  private final QuestionsListView listView;
+  private final QuestionEditView editView;
+  private final FormFactory formFactory;
+
+  @Inject
+  public QuestionController(
+      QuestionService service,
+      QuestionsListView listView,
+      QuestionEditView editView,
+      FormFactory formFactory) {
+    this.service = checkNotNull(service);
+    this.listView = checkNotNull(listView);
+    this.editView = checkNotNull(editView);
+    this.formFactory = checkNotNull(formFactory);
+  }
+
+  public CompletionStage<Result> list(String renderAs) {
+    return service
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            readOnlyService -> {
+              return ok(listView.render(readOnlyService.getAllQuestions(), renderAs));
+            });
+  }
+
+  public Result create(Request request) {
+    return ok(editView.render(request, Optional.empty()));
+  }
+
+  public CompletionStage<Result> edit(Request request, String path) {
+    return service
+        .getReadOnlyQuestionService()
+        .thenApplyAsync(
+            readOnlyService -> {
+              Optional<QuestionDefinition> definition = Optional.empty();
+              try {
+                definition = Optional.of(readOnlyService.getQuestionDefinition(path));
+              } catch (InvalidPathException e) { // If the path doesn't exist, redirect to create.
+                System.out.println(e); // TODO: What are we using for logging?
+              }
+              if (definition.isPresent()) {
+                return ok(editView.render(request, definition));
+              } else {
+                return redirect("/admin/questions/new");
+              }
+            });
+  }
+
+  public CompletionStage<Result> update(Request request) {
+    Form<QuestionForm> form = formFactory.form(QuestionForm.class);
+    QuestionForm questionForm = form.bindFromRequest(request).get();
+    try {
+      QuestionDefinition definition = questionForm.getBuilder().setId(0L).setVersion(1L).build();
```

Comment:
question: why are ID and version set to these values?

Reply:
Edit/Update isn't implemented yet, but for the questionForm to build the questionDefinition it needs and id and version. These aren't currently set in the form (we may want hidden inputs for them)  but I wanted to avoid setting the id in the `QuestionForm`

Reply:
If we are to update the exact row in db, we need the id from the existing question (hidden inputs) so eventually `QuestionForm` needs to have a `questionId` field (likely `questionVersion` as well).

For now, we no longer need to explicitly set `id` for builder to successfully build a (new) definition. Please remove it.